File: data_structure/linked_list/src/LinkedListDemo.py
# Python
# データ構造: 連結リスト (Linked List)

class NodeData:
    def __init__(self, data):
        self._data = data
        self._next = None

    def get(self):
        # 要素を取得
        return self._data

    def get_next(self):
        return self._next
    
    def set_data(self, data):
        self._data = data

    def set_next(self, next):
        self._next = next

class LinkedListData:
    def __init__(self):
        self._data = None
        self._size = 0

    def get(self):
        # 要素を取得
        return self._data

    def get_position(self, data):
        # 指定したデータの位置を検索
        if self.is_empty():
            return -1
        
        current = self._data
        position = 0
        
        while current:
            if current.get() == data:
                return position
            current = current.get_next()
            position += 1
        
        return -1

    def get_value(self, position):
        # 指定した位置の要素を取得
        if self.is_empty() or position < 0 or position >= self._size:
            print(f"ERROR: {position} は範囲外です")
            return None
        
        current = self._data
        for i in range(position):
            current = current.get_next()
        
        return current.get()

    def add(self, data, position=None):
        # 新しいノードを作成
        new_node = NodeData(data)
        
        # リストが空の場合、新しいノードをヘッドにする
        if self.is_empty():
            self._data = new_node
            self._size += 1
            return True
        
        # 位置が指定されていない場合は末尾に追加
        if position is None or position >= self._size:
            current = self._data
            # 最後のノードまで移動
            while current.get_next():
                current = current.get_next()
            current.set_next(new_node)
            self._size += 1
            return True
        
        # 先頭に追加する場合
        if position == 0:
            new_node.set_next(self._data)
            self._data = new_node
            self._size += 1
            return True
        
        # 指定した位置に挿入
        current = self._data
        for i in range(position - 1):
            current = current.get_next()
        
        new_node.set_next(current.get_next())
        current.set_next(new_node)
        self._size += 1
        return True

    def remove(self, position=None, data=None):
        # リストが空の場合
        if self.is_empty():
            print("ERROR: リストが空です")
            return False
        
        # データ指定の場合
        if data is not None:
            if self._data.get() == data:
                self._data = self._data.get_next()
                self._size -= 1
                return True
            
            current = self._data
            while current.get_next() and current.get_next().get() != data:
                current = current.get_next()
            
            if current.get_next():
                current.set_next(current.get_next().get_next())
                self._size -= 1
                return True
            else:
                print(f"ERROR: {data} は範囲外です")
                return False
        
        # 位置指定の場合
        if position is None:
            # デフォルトでは末尾要素を削除
            position = self._size - 1
        
        if position < 0 or position >= self._size:
            print(f"ERROR: {position} は範囲外です")
            return False
        
        if position == 0:
            self._data = self._data.get_next()
            self._size -= 1
            return True
        
        current = self._data
        for i in range(position - 1):
            current = current.get_next()
        
        current.set_next(current.get_next().get_next())
        self._size -= 1
        return True

    def update(self, position, data):
        # 指定した位置の要素を更新
        if self.is_empty() or position < 0 or position >= self._size:
            print(f"ERROR: {position} は範囲外です")
            return False
        
        current = self._data
        for i in range(position):
            current = current.get_next()
        
        current.set_data(data)
        return True

    def is_empty(self):
        # リストが空かどうかを返す
        return self._data is None

    def size(self):
        # リストのサイズを返す
        return self._size

    def clear(self):
        # リストを空にする
        self._data = None
        self._size = 0
        return True
    
    def display(self):
        elements = []
        current = self._data
        while current:
            elements.append(current.get())
            current = current.get_next()
        return elements

def main():
    print("LinkedList TEST -----> start")

    print("\nnew")
    linked_list_data = LinkedListData()
    print(f"  現在のデータ: {linked_list_data.display()}")

    print("\nis_empty")
    output = linked_list_data.is_empty()
    print(f"  出力値: {output}")

    print("\nsize")
    output = linked_list_data.size()
    print(f"  出力値: {output}")

    print("\nadd")
    input = 10
    print(f"  入力値: {input}")
    output = linked_list_data.add(input)
    print(f"  出力値: {output}")
    print(f"  現在のデータ: {linked_list_data.display()}")

    print("\nadd")
    input = 20
    print(f"  入力値: {input}")
    output = linked_list_data.add(input)
    print(f"  出力値: {output}")
    print(f"  現在のデータ: {linked_list_data.display()}")

    print("\nadd")
    input = (5, 0)
    print(f"  入力値: {input}")
    output = linked_list_data.add(*input)
    print(f"  出力値: {output}")
    print(f"  現在のデータ: {linked_list_data.display()}")

    print("\nadd")
    input = (15, 2)
    print(f"  入力値: {input}")
    output = linked_list_data.add(*input)
    print(f"  出力値: {output}")
    print(f"  現在のデータ: {linked_list_data.display()}")

    print("\nget_value")
    input = 1
    print(f"  入力値: {input}")
    output = linked_list_data.get_position(input)
    print(f"  出力値: {output}")

    print("\nget_value")
    input = 10
    print(f"  入力値: {input}")
    output = linked_list_data.get_position(input)
    print(f"  出力値: {output}")

    print("\nupdate")
    input = (1, 99)
    print(f"  入力値: {input}")
    output = linked_list_data.update(*input)
    print(f"  出力値: {output}")
    print(f"  現在のデータ: {linked_list_data.display()}")

    print("\nget_value")
    input = 15
    print(f"  入力値: {input}")
    output = linked_list_data.get_value(input)
    print(f"  出力値: {output}")

    print("\nget_valuefind")
    input = 100
    print(f"  入力値: {input}")
    output = linked_list_data.get_value(input)
    print(f"  出力値: {output}")

    print("\nremove")
    input = 15
    print(f"  入力値: data={input}")
    output = linked_list_data.remove(data=input)
    print(f"  出力値: {output}")
    print(f"  現在のデータ: {linked_list_data.display()}")

    print("\nremove")
    input = 0
    print(f"  入力値: position={input}")
    output = linked_list_data.remove(position=input)
    print(f"  出力値: {output}")
    print(f"  現在のデータ: {linked_list_data.display()}")

    print("\nremove")
    output = linked_list_data.remove()
    print(f"  出力値: {output}")
    print(f"  現在のデータ: {linked_list_data.display()}")

    print("\nremove")
    input = 5
    print(f"  入力値: position={input}")
    output = linked_list_data.remove(position=input)
    print(f"  出力値: {output}")
    print(f"  現在のデータ: {linked_list_data.display()}")

    print("\nclear")
    output = linked_list_data.clear()
    print(f"  出力値: {output}")
    print(f"  現在のデータ: {linked_list_data.display()}")

    print("\nis_empty")
    output = linked_list_data.is_empty()
    print(f"  出力値: {output}")

    print("\nsize")
    output = linked_list_data.size()
    print(f"出力値: {output}")

    print(f"\nremove")
    output = linked_list_data.remove()
    print(f"  出力値: {output}")
    print(f"  現在のデータ: {linked_list_data.display()}")

    print("\nLinkedList TEST <----- end")

if __name__ == "__main__":
    main()

File: data_structure/linked_list/src/LinkedListDemo.go
// Go
// データ構造: 連結リスト (Linked List)

package main

import (
	"fmt"
)

type NodeData struct {
	data interface{}
	next *NodeData
}

type LinkedListData struct {
	data *NodeData
	size int
}

func (n *NodeData) Get() interface{} {
	return n.data
}

func (n *NodeData) GetNext() *NodeData {
	return n.next
}

func (n *NodeData) SetData(data interface{}) {
	n.data = data
}

func (n *NodeData) SetNext(next *NodeData) {
	n.next = next
}

func NewLinkedListData() *LinkedListData {
	return &LinkedListData{
		data: nil,
		size: 0,
	}
}

func (l *LinkedListData) Get() *NodeData {
	return l.data
}

func (l *LinkedListData) GetPosition(data interface{}) int {
	if l.IsEmpty() {
		return -1
	}

	current := l.data
	position := 0

	for current != nil {
		if current.data == data {
			return position
		}
		current = current.next
		position++
	}

	return -1
}

func (l *LinkedListData) GetValue(position int) interface{} {
	if l.IsEmpty() || position < 0 || position >= l.size {
		fmt.Printf("ERROR: %d は範囲外です\n", position)
		return nil
	}

	current := l.data
	for i := 0; i < position; i++ {
		current = current.next
	}

	return current.data
}

func (l *LinkedListData) Add(data interface{}, position ...int) bool {
	newNode := &NodeData{data: data}

	if l.IsEmpty() {
		l.data = newNode
		l.size++
		return true
	}

	pos := -1
	if len(position) > 0 {
		pos = position[0]
	}

	if pos == -1 || pos >= l.size {
		current := l.data
		for current.next != nil {
			current = current.next
		}
		current.next = newNode
		l.size++
		return true
	}

	if pos == 0 {
		newNode.next = l.data
		l.data = newNode
		l.size++
		return true
	}

	current := l.data
	for i := 0; i < pos-1; i++ {
		current = current.next
	}

	newNode.next = current.next
	current.next = newNode
	l.size++
	return true
}

func (l *LinkedListData) Remove(args ...interface{}) bool {
	if l.IsEmpty() {
		fmt.Println("ERROR: リストが空です")
		return false
	}

	if len(args) > 0 {
		switch v := args[0].(type) {
		case int:
			return l.removeByPosition(v)
		default:
			return l.removeByData(v)
		}
	}

	return l.removeByPosition(l.size - 1)
}

func (l *LinkedListData) removeByData(data interface{}) bool {
	if l.data.data == data {
		l.data = l.data.next
		l.size--
		return true
	}

	current := l.data
	for current.next != nil && current.next.data != data {
		current = current.next
	}

	if current.next != nil {
		current.next = current.next.next
		l.size--
		return true
	}

	fmt.Printf("ERROR: %v は範囲外です\n", data)
	return false
}

func (l *LinkedListData) removeByPosition(position int) bool {
	if position < 0 || position >= l.size {
		fmt.Printf("ERROR: %d は範囲外です\n", position)
		return false
	}

	if position == 0 {
		l.data = l.data.next
		l.size--
		return true
	}

	current := l.data
	for i := 0; i < position-1; i++ {
		current = current.next
	}

	current.next = current.next.next
	l.size--
	return true
}

func (l *LinkedListData) Update(position int, data interface{}) bool {
	if l.IsEmpty() || position < 0 || position >= l.size {
		fmt.Printf("ERROR: %d は範囲外です\n", position)
		return false
	}

	current := l.data
	for i := 0; i < position; i++ {
		current = current.next
	}

	current.data = data
	return true
}

func (l *LinkedListData) IsEmpty() bool {
	return l.data == nil
}

func (l *LinkedListData) Size() int {
	return l.size
}

func (l *LinkedListData) Clear() bool {
	l.data = nil
	l.size = 0
	return true
}

func (l *LinkedListData) Display() []interface{} {
	elements := []interface{}{}
	current := l.data
	for current != nil {
		elements = append(elements, current.data)
		current = current.next
	}
	return elements
}

func main() {
	fmt.Println("LinkedList TEST -----> start")

	fmt.Println("\nnew")
	linkedListData := NewLinkedListData()
	fmt.Printf("  現在のデータ: %v\n", linkedListData.Display())

	fmt.Println("\nis_empty")
	output := linkedListData.IsEmpty()
	fmt.Printf("  出力値: %v\n", output)

	fmt.Println("\nsize")
	sizeOutput := linkedListData.Size()
	fmt.Printf("  出力値: %d\n", sizeOutput)

	fmt.Println("\nadd")
	input := 10
	fmt.Printf("  入力値: %d\n", input)
	addOutput := linkedListData.Add(input)
	fmt.Printf("  出力値: %v\n", addOutput)
	fmt.Printf("  現在のデータ: %v\n", linkedListData.Display())

	fmt.Println("\nadd")
	input = 20
	fmt.Printf("  入力値: %d\n", input)
	addOutput = linkedListData.Add(input)
	fmt.Printf("  出力値: %v\n", addOutput)
	fmt.Printf("  現在のデータ: %v\n", linkedListData.Display())

	fmt.Println("\nadd")
	input = 5
	inputPos := 0
	fmt.Printf("  入力値: (%d, %d)\n", input, inputPos)
	addOutput = linkedListData.Add(input, inputPos)
	fmt.Printf("  出力値: %v\n", addOutput)
	fmt.Printf("  現在のデータ: %v\n", linkedListData.Display())

	fmt.Println("\nadd")
	input = 15
	inputPos = 2
	fmt.Printf("  入力値: (%d, %d)\n", input, inputPos)
	addOutput = linkedListData.Add(input, inputPos)
	fmt.Printf("  出力値: %v\n", addOutput)
	fmt.Printf("  現在のデータ: %v\n", linkedListData.Display())

	fmt.Println("\nget_value")
	input = 1
	fmt.Printf("  入力値: %d\n", input)
	outputPos := linkedListData.GetPosition(input)
	fmt.Printf("  出力値: %d\n", outputPos)

	fmt.Println("\nget_value")
	input = 10
	fmt.Printf("  入力値: %d\n", input)
	outputPos = linkedListData.GetPosition(input)
	fmt.Printf("  出力値: %d\n", outputPos)

	fmt.Println("\nupdate")
	inputPos = 1
	input = 99
	fmt.Printf("  入力値: (%d, %d)\n", inputPos, input)
	updateOutput := linkedListData.Update(inputPos, input)
	fmt.Printf("  出力値: %v\n", updateOutput)
	fmt.Printf("  現在のデータ: %v\n", linkedListData.Display())

	fmt.Println("\nget_value")
	input = 15
	fmt.Printf("  入力値: %d\n", input)
	getValueOutput := linkedListData.GetValue(input)
	fmt.Printf("  出力値: %v\n", getValueOutput)

	fmt.Println("\nget_valuefind")
	input = 100
	fmt.Printf("  入力値: %d\n", input)
	getValueOutput = linkedListData.GetValue(input)
	fmt.Printf("  出力値: %v\n", getValueOutput)

	fmt.Println("\nremove")
	input = 15
	fmt.Printf("  入力値: data=%d\n", input)
	removeOutput := linkedListData.Remove(input)
	fmt.Printf("  出力値: %v\n", removeOutput)
	fmt.Printf("  現在のデータ: %v\n", linkedListData.Display())

	fmt.Println("\nremove")
	inputPos = 0
	fmt.Printf("  入力値: position=%d\n", inputPos)
	removeOutput = linkedListData.Remove(inputPos)
	fmt.Printf("  出力値: %v\n", removeOutput)
	fmt.Printf("  現在のデータ: %v\n", linkedListData.Display())

	fmt.Println("\nremove")
	removeOutput = linkedListData.Remove()
	fmt.Printf("  出力値: %v\n", removeOutput)
	fmt.Printf("  現在のデータ: %v\n", linkedListData.Display())

	fmt.Println("\nremove")
	inputPos = 5
	fmt.Printf("  入力値: position=%d\n", inputPos)
	removeOutput = linkedListData.Remove(inputPos)
	fmt.Printf("  出力値: %v\n", removeOutput)
	fmt.Printf("  現在のデータ: %v\n", linkedListData.Display())

	fmt.Println("\nclear")
	clearOutput := linkedListData.Clear()
	fmt.Printf("  出力値: %v\n", clearOutput)
	fmt.Printf("  現在のデータ: %v\n", linkedListData.Display())

	fmt.Println("\nis_empty")
	output = linkedListData.IsEmpty()
	fmt.Printf("  出力値: %v\n", output)

	fmt.Println("\nsize")
	sizeOutput = linkedListData.Size()
	fmt.Printf("出力値: %d\n", sizeOutput)

	fmt.Println("\nremove")
	removeOutput = linkedListData.Remove()
	fmt.Printf("  出力値: %v\n", removeOutput)
	fmt.Printf("  現在のデータ: %v\n", linkedListData.Display())

	fmt.Println("\nLinkedList TEST <----- end")
}
